### DISCUSSION TITLE
Strange behavior of Spacy french tokenizer

### DISCUSSION THREAD

**Discussion by Pandalei97** [explosion/spaCy#8316] [Help: Coding & Implementations] [answered]
[Update]
I have solved this problem. Indeed 'Marie' exist in the `_hyphen_prefix`. Sorry to bother you, you can delete this thread (I can't do it by myself).

But still, I have to say that the some default hyphen_prefix are strange. If other users are having the same problem, you may consider using your own 'token-match' by slightly modifying the default one.

--------------------------------------------------
Hello ! I want to add some rules to the Spacy French tokenizer to manage hyphens. But I'm having trouble with several words, which doesn't do what I expected. 

My custom tokenizer:
```python
def custom_tokenizer(nlp):
    custom_prefixes = r'^-'
    custom_infixes = r'-'
    custom_suffixes = r'[-\+]$'

    infix_re = compile_infix_regex(tuple(list(nlp.Defaults.infixes) + [custom_infixes]))
    prefix_re = compile_prefix_regex(tuple(list(nlp.Defaults.prefixes) + [custom_prefixes]))
    suffix_re = compile_suffix_regex(tuple(list(nlp.Defaults.suffixes) + [custom_suffixes]))

    return Tokenizer(nlp.vocab,
                     nlp.Defaults.tokenizer_exceptions,
                     prefix_search=prefix_re.search,
                     infix_finditer=infix_re.finditer,
                     suffix_search=suffix_re.search,
                     token_match=nlp.tokenizer.token_match)
```
For the text : `st-marie- Marie-May- Marie-Mahtddf- Mari-grtt- tokenisation-`.

I expect to have:
`[st, -, marie, -, Marie, -, May, -, Marie, -, Mahtddf, -, Mari, -, grtt, -, tokenisation, -]`

But I got actually:
`[st, -, marie, -, Marie-May-, Marie-Mahtddf-, Mari, -, grtt, -, tokenisation, -]`

Apparently it's due to the language default token_match. By not adding `nlp.tokenizer.token_match`, this isn't causing the problem. However, I can't find any exception pattern in Spacy's tokenizer_exceptions.py that should match the case with 'Marie-....'.

Is that normal to have this result ? This can be potentially causing problem with our matcher patterns for some other words.

My spacy version: 3.0.6

**Comment by polm** [comment, marked as answer]
Glad you solved this, please note we don't generally delete threads. This might be useful to someone in the future.

As a general point of advice about the tokenizer, if you don't understand a result you can use `tokenizer.explain(text)` to get an explanation of where each token came from. So for your `Marie-` examples it gave `TOKEN_MATCH`, which makes it clear what's going on there. 

(I understand this may be why you figured out it was coming from `token_match`, but I want to make it clear for future readers.)
